### DISCUSSION TITLE
Cast to TV

### DISCUSSION THREAD

**Discussion by Bnyro** [libre-tube/LibreTube#3493] [Ideas]
### Describe your suggested feature

Option to cast to a TV (e.g. Chromecast or FireTv Stick)

### Other details

I can't add this because I don't have the needed resources for it.

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I will fill out all of the requested information in this form.

**Comment by FireMasterK** [comment]
https://developers.google.com/cast/docs/android_sender

It looks like you require a proprietary library :/ for doing it officially

But, it looks like we have a microG variant! https://github.com/microg/GmsCore/tree/master/play-services-cast-api

**Comment by luantak** [comment]
I think it's okay to use the official api because it links against the library provided by the os i.e. if you don't have Google Play Services casting will not work for you, or if you have replaced Google play services with microg it will use that instead.

**Comment by FireMasterK** [comment]
It's proprietary code, and would not be allowed in f-droid. I would rather suggest using the microg api, which is open source.

**Comment by the4anoni** [comment]
Remember that Chromecast uses Cast API, while other devices use pairing by code (YouTube.com/activate). Second option provides support for much more devices (Tizen, WebOS TVs, SmartTubeNext, official Youtube app on Fire TV possibly too, etc.).

SmartTubeNext is open source alternative client independent on gapps that works on any Android device meant for TV (doesn't matter is this official Android TV, non certified Chinese Android Box, Fire TV, etc.).

I think pairing by code should be higher on priority list than casting, also it looks like easier to implement.

**Comment by ghost** [comment]
This can be achieved easily without any development with the current version. If you want to cast a video, just press the "Open..." button right beside the Download button and cast the video with google photos video player.

**Comment by the4anoni** [comment]
> This can be achieved easily without any development with the current version. If you want to cast a video, just press the "Open..." button right beside the Download button and cast the video with google photos video player.

I don't have installed Google Photos, also It would be nice if we could have button in app for that.

**Comment by ghost** [comment]
> > This can be achieved easily without any development with the current version. If you want to cast a video, just press the "Open..." button right beside the Download button and cast the video with google photos video player.
> 
> I don't have installed Google Photos, also It would be nice if we could have button in app for that.

Yeah, that would be nice too, I just want to share an alternative way.

**Comment by ghost** [comment]
Assign me to issue

**Comment by Silther** [comment]
> This can be achieved easily without any development with the current version. If you want to cast a video, just press the "Open..." button right beside the Download button and cast the video with google photos video player.

The button was removed in the last update.

**Comment by Bnyro** [comment]
It might have a comeback with #2292 though, didn't decide it yet.

**Comment by youdontneedtoknow22** [comment]
Skytube solved this problem with having 2 Versions, one is a FOSS version on F-Droid without the cast feature, and an extra version hosted on Github with Chromecast.

https://github.com/SkyTubeTeam/SkyTube/issues/22

**Comment by Silther** [comment]
Maybe I'm an optimist, but I just hope that Matter TV will catch on and that it will be open source. but it won't be released until the end of this year at the earliest.
I would rather prefer the standard if it is ever made available.
https://www.theverge.com/2021/12/9/22824559/matter-tv-streaming-devices-smart-home-casting-protocol-support

**Comment by the4anoni** [comment]
> 
![standards](https://user-images.githubusercontent.com/13403084/229208195-b5108691-43e8-4118-b2e0-92e32890b46d.png)

**Comment by jaredsburrows** [comment]
You can with NewPipe + VLC. See https://github.com/TeamNewPipe/NewPipe/issues/668#issuecomment-959954722.

Possible we can do something like that here?

**Reply by Silther** [reply]
> Possible we can do something like that here?

You can already use an external player, just long press the share button.

**Comment by jaredsburrows** [comment]
@xIUPITERx  Thanks.

**Comment by Niko4u5** [comment]
https://github.com/MarcoLucidi01/ytcast#thanks I found this tool to connect to the TV app which links some resources on how it works

**Comment by SecularSteve** [comment]
https://github.com/libre-tube/LibreTube/issues/4637

**Comment by Eco-Gaming** [comment]
Another option for casting would be using FCast:
https://fcast.org/
https://gitlab.futo.org/videostreaming/fcast

It's a source available (~~might not be foss~~) chromecast alternative. FCast is used in FUTOs client for Youtube, etc. ([GrayJay](https://grayjay.app/) ([GitLab](https://gitlab.futo.org/videostreaming/grayjay))) and according to the devs it's supposed to be easy to implement.
I tried FCast with GrayJay and it worked perfectly, but I prefer Libretube over GrayJay so it would be great to have FCast in Libretube too.

Edit: FCast uses the MIT license

**Comment by Silther** [comment]
Matter Casting finally seems to be ready, but I didn't yet find documantation how to implement it.

**Comment by Savasdotexe** [comment]
Any update to this? YouTube ReVanced are able to do this, if it's possible with microG, I'm really hoping this gets some attention.

The one thing stopping me from actually completely uninstalling YouTube is the inability to hit the cast option on this. It's basically how I watch things (find something I like on my phone and press cast to the TV). 

I guess on newer model TV's and for those with more easier remotes that pretty much behave like mouses you skip using your phone and browse via the TV, however that ends up with having to look up things you spot on your hand device to play or having duplicate subscription and playlists. Not ideal.

Matter TV looks promising, however I doubt my TV will be compatible, no doubt they'll want you to login, so I think I prefer simply casting. there's still the issue with maintaining duplicate lists with Matter?

**Comment by curiousercreative** [comment]
FWIW, VLC on F-Droid has Chromecast support OOB for me on /e/os (microg)

**Comment by Savanta** [comment]
So I made a fork. I should start with saying that I don't have any experience with Kotlin, LoungeAPI is not well documented and I made it with help from copilot. I kept prompting Gemini for documentation on LoungeAPI and maybe, maybe broke out of it's jail so it spit out documentation on how to use it.
I also used https://github.com/FabioGNR/pyytlounge and https://github.com/enriquecidok/youtube-lounge-api
What works: syncing progress both ways, play/pause, changing video both ways, pulling video from remote on initial connect is wobbly but mostly works, heartbeat should work but can be wobbly when app is in background. 
What doesn't work: audio only, you tell me
Here is it is: https://github.com/Savanta/LibreTube
I should warn you... I don't have experience with Kotlin so code could be at best of dubious quality.
Later today I will release apk.
